Added retry on DuplicateKeyError

AbhijithSivarajan opened the pull request:
Added support to retry once more on occurrence of `DuplicateKeyError` during mongodb `replace_one` operation.

AbhijithSivarajan committed:
Added retry on DuplicateKeyError for pymongo replace_one operation

There is a rare scenario where two threads may attempt the
replace/update operation at the same time.
Both threads will not find any match based on the query and
will try to insert the data.
One thread might update the data whereas
the other one will receive a DuplicateKeyError
because of the inserted data.
So to avoid this issue,
added a retry on occurance of DuplicateKeyError.
Adding a retry will return the previously inserted data
based on the query and help update the data.
Reference: https://jira.mongodb.org/browse/SERVER-14322

AbhijithSivarajan committed:
Removed handling of global exceptions

Any global exception will be raised immediately as it is,
so the developer will have an idea and stacktrace
of any unexpected exception

AbhijithSivarajan committed:
Moved the asynctest package to test_packages in ubuild.py file

toumorokoshi commented on tycho/db/event/__init__.py:
```diff
@@ -1,14 +1,20 @@
 
+
 MAX_RECURSIVE_DEPTH = 100
+WAIT_TIME_S = 1
```
can we expand this variable name? there's not much valuable in abbreviating it to "s". It's not the most ambiguous, but "WAIT_TIME_SECONDS" is definitely clearer.

AbhijithSivarajan replied:
Updated the variable name to `WAIT_TIME_SECONDS`

toumorokoshi commented on tycho/db/event/__init__.py:
```diff
@@ -43,9 +49,18 @@ def __init__(self, collection):
-        return result
+
+        retries = 1
+        while retries >= 0:
```
this can result in an infinite loop if the exception raised is not DuplicateKeyError, since the retries don't decrement there.

How about a global catch-all that decrements retries instead?

AbhijithSivarajan replied:
The occurrence of any exception other than DuplicateKeyError will directly be raised, thereby breaking the loop and returning the control to the callee. So, it might not go into an infinite loop. Do let me know if I have got it right.

toumorokoshi replied:
ah, good point.

toumorokoshi replied:
Ok, that makes sense. There's probably a more robust strategy here (like CAS) https://stackoverflow.com/questions/11991167/in-mongo-any-way-to-do-check-and-setting-like-atomic-operation

I think you can do this by running replace_one with the full old data, running replace_one, then retrying if it doesn't match. That would in general be more robust, but I don't see why this wouldn't work.

AbhijithSivarajan replied:
Used `find_one_and_replace` operation instead of `replace_one`

toumorokoshi replied:
I don't think this helps, as the issue notes that findAndModify is also affected, in the case that upsert is True:

https://jira.mongodb.org/browse/SERVER-14322

But again, I think this retry works well in most cases, and the variable timeout helps with worst case robustness.

toumorokoshi reviewed (changes requested):
It's not really clear from any of the commit messages what the motivation for this change is.

Even if we do use a separate ticket management system, we should make sure that the source code history contains good context on why we're making the changes we are. 

Here's my favorite post around commit messages: https://chris.beams.io/posts/git-commit/

That aside, I don't think there's a good scenario where the handling for DuplicateKeyError is to retry. DuplicateKeyErrors are raised because there's an attempt to add a record in a collection with the same key. It's not clear to me why just retrying resolves the issue.

Final note is that the current retry logic can cause an infinite loop, so we should avoid that.

AbhijithSivarajan committed:
Resolved merge conflict with master branch

AbhijithSivarajan committed:
Renamed WAIT_TIME_S variable to WAIT_TIME_SECONDS

AbhijithSivarajan committed:
Removed some unused imports

AbhijithSivarajan commented:
> It's not really clear from any of the commit messages what the motivation for this change is.
> 
> Even if we do use a separate ticket management system, we should make sure that the source code history contains good context on why we're making the changes we are.
> 
> Here's my favorite post around commit messages: https://chris.beams.io/posts/git-commit/
> 
> That aside, I don't think there's a good scenario where the handling for DuplicateKeyError is to retry. DuplicateKeyErrors are raised because there's an attempt to add a record in a collection with the same key. It's not clear to me why just retrying resolves the issue.
> 
> Final note is that the current retry logic can cause an infinite loop, so we should avoid that.


Thanks for the comment and for sharing the doc. I have updated the commit messages to be more descriptive.

Regarding the issue, there is a rare scenario where two threads may attempt the replace/update operation at the same time.
Both threads will not find any match based on the query and will try to insert the data.
One thread might update the data whereas the other one will receive a DuplicateKeyError because of the inserted data.
So to avoid this issue, I have added a retry on the occurrence of DuplicateKeyError.
Adding a retry will return the previously inserted data based on the query and help update the data.
Reference: https://jira.mongodb.org/browse/SERVER-14322

toumorokoshi commented on tycho/db/event/__init__.py:
```diff
@@ -43,9 +48,18 @@ def __init__(self, collection):
+                    {"_id": id}, new_data, upsert=insert)
+                return result
+            except DuplicateKeyError:
+                await asyncio.sleep(WAIT_TIME_SECONDS)
```
should this be a random wait? I guess it's unusual for 3 operations to be identical, but in that situation we have both remaining retries occurring within one second. It would be better to be random between 0 to WAIT_TIME_SECONDS.

AbhijithSivarajan replied:
Updated the wait time to be a random value between 1 to 5 seconds.

toumorokoshi replied:
The wait time could be much, much lower. most replace_one operations take on the order of 10-20ms. I would have said make the wait variable between 0 and 500ms.

AbhijithSivarajan replied:
My Bad. Thanks for clarifying. Updated the wait to be random between 0 and 500ms.

toumorokoshi reviewed (approved):
I think check-and-set is a better strategy, but this code should be fine in any practical scenario I can think of. Thanks!

AbhijithSivarajan committed:
Used atomic find_one_and_replace operation instead of replace_one

Atomic operation are more robust when dealing
with threads and write operations.

Also, updated the sleep time on retry to be
a random value between 1 to 5,
so that different threads will retry the operation after
different sleep times

AbhijithSivarajan committed:
Updated the sleep time to be random value between 0 and 500ms

The db operations usually take around 10-20ms.
So, updated the sleep time before a retry
to be between 0 and 500ms.
